### ISSUE TITLE
Self Supervised Loss

### ISSUE THREAD

**Issue by VictorZuanazzi** [DylanWusee/PointPWC#1]
Thank you for making the code available =)

I want to reproduce the Self-Supervised results from the paper. However, this part is not available yet. I coded the losses myself, but I am not getting the same results.

The `ball_query` and `grouping_operation` are the ones in the pointnet2 package.

```
def multiScaleSelfSupervisedLoss(flows, pc1, pc2, alpha=[0.02, 0.04, 0.08, 0.16], beta=[1., 1., .3]):
    """self supervised losses as described in PointPWC paper"""
    
    neighbors = 8
    loss = 0
    for i, (flow, c1, pc_target) in enumerate(zip(flows, pc1, pc2)):
        # making data CUDA friendly
        pc_target = pc_target.contiguous()
        pc_pred = (c1 + flow).contiguous()
        flow = flow.contiguous()
        pc_pred_t = pc_pred.transpose(2, 1).clone().detach()
        pc_target_t = pc_target.transpose(2, 1).clone().detach()

        # chamfer distace:
        idxs_t = ball_query(1.0, 1, pc_target_t, pc_pred_t)  # idxs of nearest neighbors of pred in target
        knn = grouping_operation(pc_target, idxs_t)  # retrieve nearest neighbors
        dist1 = (knn.squeeze() - pc_pred).norm(dim=1).sum(dim=-1)  # distances from pred to target

        idxs = ball_query(1.0, 1, pc_pred_t, pc_target_t)  # idxs of nn of target in pred
        knn = grouping_operation(pc_pred, idxs)  # retrieve nn
        dist2 = (knn.squeeze() - pc_target).norm(dim=1).sum(dim=-1)  # distances from target to pred
        dist = (dist1 + dist2)  # chamfer distance

        # local smootheness
        idxs = ball_query(1.0, neighbors, pc_pred_t, pc_pred_t)  # idxs of nn of pred in pred
        flow_local = grouping_operation(flow, idxs)  # retrieve flow of nn
        local_consistency_loss = (flow.unsqueeze(3) - flow_local).norm(dim=1).mean(dim=-1).sum( dim=-1)  # nn flow consistency

        # laplacian constraint
        knn_p = grouping_operation(pc_pred, idxs)  # uses the same idxs already calculated from local smoothness
        lp = (pc_pred.unsqueeze(3) - knn_p).mean(dim=-1)  # laplacian vector as described in the paper

        idxs = ball_query(1.0, neighbors, pc_target_t, pc_target_t)  # idxs of nn from target in target
        knn_t = grouping_operation(pc_target, idxs)  # retrieve nn
        lt = (pc_target.unsqueeze(3) - knn_t).mean(dim=-1)  # laplacian vector as described in the paper
        knn_l = grouping_operation(lt, idxs_t)  # uses idxs calculated fro chamfer dist from pred to taget
        laplace_error = (lp - knn_l.squeeze()).norm(dim=1).sum(dim=-1)  # laplacian constraint

        loss += alpha[i] * (beta[0] * dist + beta[1] * local_consistency_loss + beta[2] * laplace_error).mean()

    return loss
```

Is there something I am missing in the implementation?

**Comment by DylanWusee** [comment]
Hi Victor,

I think the implementation is generally good. Not sure what is wrong on your side at this moment. 

Could you please post some of your training logs here? And your training setup, such as batch size, etc.

We will release the supervised lose code when the code is ready.

**Comment by VictorZuanazzi** [comment]
The eval EPE increases instead of decreasing in the first iterations. It stays in between 2.5 and 3.5 in for as long as I have trained so far (around 60 epochs).

The last loss logs:
```
EPOCH 61 eval mean epe3d: 4.493206  mean eval loss: 1734.885864
Best epe loss is: 2.12085
Learning rate:0.001000

100%|██████████| 313/313 [06:37<00:00,  1.27s/it]
EPOCH 62 train mean loss: 4551.929688
100%|██████████| 126/126 [01:47<00:00,  1.17it/s]
EPOCH 62 eval mean epe3d: 4.013077  mean eval loss: 1633.128052
Best epe loss is: 2.12085
Learning rate:0.001000

100%|██████████| 313/313 [06:36<00:00,  1.27s/it]
EPOCH 63 train mean loss: 4527.550293
100%|██████████| 126/126 [01:47<00:00,  1.17it/s]
EPOCH 63 eval mean epe3d: 4.381819  mean eval loss: 1737.572510
Best epe loss is: 2.12085
Learning rate:0.001000

100%|██████████| 313/313 [06:37<00:00,  1.27s/it]
EPOCH 64 train mean loss: 4506.511719
100%|██████████| 126/126 [01:47<00:00,  1.17it/s]
EPOCH 64 eval mean epe3d: 4.718708  mean eval loss: 1928.377930
Best epe loss is: 2.12085
```

The set up is the same as the one in `config_train.yaml`, I just changed the bach size to 16 (instead of 8) for better use of GPU:

```
#gpu: "0"
multi_gpu: null
model_name: PointConv

full: False

arch: PointConv
last_relu: False
allow_less_points: False

# general args
batch_size: 16 # 8
epochs: 800
learning_rate: 0.001

optimizer: Adam
pretrain: null #./experiment/PointConvFlyingthings3d-2019-11-05_12-58/checkpoints/PointConv_164_0.0842.pth
weight_decay: 0.0001
supervised: False

dim: 3
num_points: 8192

DEVICE: cuda

evaluate: False

dataset: FlyingThings3DSubset 
data_root: ./flyingthings3d

data_process:
  DEPTH_THRESHOLD: 35.
  NO_CORR: True

# data augmentation
aug_together:
  degree_range: 0.1745329252
  shift_range: 1.
  scale_low: 0.95
  scale_high: 1.05
  jitter_sigma: 0.01
  jitter_clip: 0.00

aug_pc2:
  degree_range: 0.
  shift_range: 0.3
  jitter_sigma: 0.01
  jitter_clip: 0.00

# miscalleneous
workers: 16

```

**Comment by VictorZuanazzi** [comment]
What is the size you set for the neighborhood?
I am using `radius = 1.0` as default and `k = 8` for the knn. Maybe those parameters have a large impact on the effectiveness of the losses.

**Comment by DylanWusee** [comment]
The radius is way too large. And, I am not using the ball_query since it does not give the real K nearest neighbor. Try to compute the squared distance and use torch.topk instead.

**Comment by FangyingZhai** [comment]
Did you test the time to calculate the total loss/iteration if the batch size = 8?

**Comment by ashesknight** [comment]
I would like to ask if you can open the source code of the self-supervised part, because the results of my own replication are far from those of your release on ECCV2020.

**Comment by baurst** [comment]
@DylanWusee Thank you for publishing the code.

@ashesknight May I ask how far off your results are from the results in the paper? I'm trying to find out whether this work is suitable as a baseline.

**Comment by DylanWusee** [comment]
I added a branch self_loss. It includes a naive implementation of self_supervised loss there, which is the "multiScaleChamferSmoothCurvature".

**Comment by DylanWusee** [comment]
I will merge it to the master branch shortly. With the multiScaleChamferSmoothCurvature loss, the model should easily reach the results in the paper(0.1213 for FlyingThings3D, most of the time even better).

**Comment by DylanWusee** [comment]
Code is merged to master.
